text/0000-reads-clauses-on-methods.md

```diff
@@ -0,0 +1,200 @@
+
+```dafny
+
+class FibonacciMemoizer {
+  var cache: map<nat, nat>
+  constructor() reads {} {
+    cache := map[];
+  }
+  method Get(n: nat) returns (r: nat) 
+    reads this
+    modifies this
+  {
+    if n <= 2 {
+      return 1;
+    }
+    if n in cache {
+      r := cache[n];
+    } else {
+      var oneBefore := Get(n - 1);
+      var twoBefore := Get(n - 2);
+      r := oneBefore + twoBefore;
+      cache := cache[n := r];
+    }
+  }
+}
+
+// Sharing a memoizer instance between threads is not allowed.
+method {:extern} UnsafeFibonacci(n: nat, memoizer: FibonacciMemoizer) returns (r: nat)
+  reads {}
+{
+  r := memoizer.Get(n);   // Error: insufficient reads clause to invoke method
+                          // Error: call may violate context's modifies clause
+}
+
+// It IS allowed, though, to manipulate mutable state that is freshly
+// allocated by a single thread. This does not violate the reads or modifies clauses.
+method {:extern} SafeFirstNFibonaccis(n: nat) returns (rs: seq<nat>)
+  reads {}
+{
+  var memoizer := new FibonacciMemoizer();
+  rs := [];
+  for i := 1 to n + 1 {
+    var r := memoizer.Get(i);
+    rs := rs + [r];
+  }
+}
+```
+
+This is most effective when the person providing the specification is different from the person implementing it. It ensures that the implementation cannot interact with shared state by accident, especially deep within shared libraries that are perfectly safe to use in a single-threaded context.
+
+When shared state is necessary, it must be implemented externally in native code, where native facilities for synchronizing access mutable state can be used.
+
+```dafny
+datatype Option<T> = Some(value: T) | None
+
+//
+// Wrapper around a concurrent data structure, such as
+// ConcurrentHashMap<K, V> in Java.
+//
+// Note the lack of a ghost variable to track the state of the map,
+// which intentionally puts the mutable state outside of the Dafny-verified model.
+// These methods instead appear to have non-deterministic behavior to Dafny, which accurately
+// represents the fact that it cannot assume anything about state that might be changed by
+// other threads.
+//
+class {:extern} ExternalMutableMap<K, V> {
+  constructor {:extern} () reads {}
+  method {:extern} Put(k: K, v: V) reads {}
+  method {:extern} Get(k: V) returns (v: Option<V>) reads {}
+}
+
+class FibonacciMemoizer {
+  const cache: ExternalMutableMap<nat, nat>
+  constructor() reads {} {
+    cache := new ExternalMutableMap();
+  }
+  method Get(n: nat) returns (r: nat)
+    reads {}
+  {
+    if n <= 2 {
+      return 1;
+    }
+    var cached := cache.Get(n);
```

Comment:
That's right, `reads` and `modifies` only care about mutable fields, so `var`s are in scope but not `const`s. I agree that's a bit surprising but already how `reads` works on functions.

Sure, we don't "know" if `ExternalMutableMap` is mutable since the verifier has no visibility into the external code. But there's no way around that, and modelling the assumptions we might make about external code using ghost state as in https://github.com/dafny-lang/dafny/wiki/Modeling-External-State-Correctly breaks down if that external state can change without Dafny being aware of it. The idea here is to verify that a method does not interact with any mutable state on the _Dafny_ heap, but may still interact with "the outside world", and it's up to the external code to be thread-safe when necessary.

Reply:
And if `ExternalMutableMap` was not implemented externally, how would you know that it's immutable?

Reply:
Discussed this offline to understand the question better, and I can clarify how this all works with a few more details.

Note that clauses such as `reads` and `modifies` are not recursive. Take these classes as an example:

```dafny
class A {
  constructor() {}
  var foo: int
  var b: B
}
class B {
  var bar: int
}
```

If you had an object `a: A`, the clause `reads {a}` would only allow you to reference `a.foo`, but NOT `a.b.foo`. For the latter to be allowed, you would need `reads {a, a.b}`. This is why the `Valid() / Repr` idiom described in Chapters 8 and 9 of [Using Dafny, an Automatic Program Verifier](http://leino.science/papers/krml221.pdf) is useful for tracking what areas of the heap your code needs to read and write, without depending on the internals of your components.

That means Dafny doesn't allow the `cache.Get` call because it believes `ExternalMutableMap ` is immutable, it's allowed because the `cache.Get` method has `reads {}`, which means the method does not read any state in the Dafny heap.

Reply:
Btw, I presume that `reads` clauses are subject to themselves. That is, a `reads` clause has to be "closed", in the same way that they are for functions. Their well-formedness should then also be allowed to assume the method precondition, just as for functions. (The Dafny verifier has an encoding for this kind of thing. Look for the parameter called `delayedReadsChecks` or something like that.)